README.md
<h1 align="center">Welcome to dye.js 👋</h1>
<p>
  <img alt="Version" src="https://img.shields.io/badge/version-0.4-blue.svg?cacheSeconds=2592000" />
  <img src="https://img.shields.io/badge/npm-%3E%3D8.19.2-blue.svg" />
  <img src="https://img.shields.io/badge/node-%3E%3D16.17.0-blue.svg" />
  <a href="https://github.com/Slaveworx/dye.js#readme" target="_blank">
    <img alt="Documentation" src="https://img.shields.io/badge/documentation-yes-brightgreen.svg" />
  </a>
  <a href="https://github.com/Slaveworx/dye.js/graphs/commit-activity" target="_blank">
    <img alt="Maintenance" src="https://img.shields.io/badge/Maintained%3F-yes-green.svg" />
  </a>
  <a href="https://github.com/Slaveworx/dye.js/stable/LICENSE" target="_blank">
    <img alt="License: MIT" src="https://img.shields.io/github/license/Slaveworx/dye.js/stable/" />
  </a>
</p>

> Tiny and Blazing Fast Library that changes HTML elements' colors based on their self or parent background color.

### 🏠 [Homepage](https://deviago.me)

## Prerequisites

- npm >=8.19.2
- node >=16.17.0
- Works in any modern browser

## Usage

### ELEMENTS

Just apply the property "dye" to any regular HTML element and see the magic happen.
Please note that the `dye` attribute by itself _will only work with elements that have a background color themselves_.

```html
<p dye>This element will change color depending on self background color</p>
```

The instruction above will only work on element which have a background color themselves or in images in which the parent element has a background color.

If that is not the case, then you need to specify the parent:

```html
<p dye=".this-is-the-element-with-bg">
  This element will change color depending on the provided element's background
  color
</p>
```

### IMAGES

At the moment, this library only changes the color of black / white images or icons.

So you just follow the same logic as above, but there is a small catch: if your source image is originally black, you need to specify another parameter (`blackimg`) to the element:

```html
<img src="blackimage.png" dye="#this-is-an-id" blackimg>This element will change color depending on the parent background color</img>
```

Note: by default, when the element is an image, the library will always check the parent background color. If you wish to use another color in a remote relative, just provide the element to the `dye=""` parameter.

```html
<img src="blackimage.png" dye="#this-in-the-element-with-bg-color" blackimg>This element will change color depending on the parent background color</img>
```

### Gradients

From version 0.4.3 dye.js now supports gradients up to three colors.

Please bear in mind that depending on gradient direction or gradient transparencies and color positions, the output might not be 100% accurate.
However, dye.js works very well with relatively simple gradient backgrounds.
Note: this feature only works changing text color for now.
Note 2: This feature supports colors in RGB, RGBA, and HEX.

To use it you just need to add the `gradient` attribute to elements that have a gradient background and that you wish to change color dynamically.

```html
<p dye="#this-in-the-parent-id-with-gradient-bg" gradient>
  This element will change color depending on the parent gradient background.
</p>
```

### IMPORTANT NOTES

When you specify a selector, you need to identify it as a `class` or `id`, so please use the below syntax:

```html
<img src="blackimage.png" dye="#this-is-an-id" blackimg>This element will change color depending on the parent background color</img>
```

```html
<img src="blackimage.png" dye=".this-is-an-class" blackimg>This element will change color depending on the parent background color</img>
```

Note that we use `.` or `#` to let dye.js know that the selector is and `id` or `class`.

## Author

👤 **Tiago M. Galvão**

- Website: http://deviago.me
- Github: [@Slaveworx](https://github.com/Slaveworx)

## 🤝 Contributing

Contributions, issues and feature requests are welcome!<br />Feel free to check [issues page](https://github.com/Slaveworx/dye.js/issues).

## Show your support

Give a ⭐️ if this project helped you!

## 📝 License

Copyright © 2023 [Tiago M. Galvão](https://github.com/Slaveworx).<br />
This project is [MIT](https://github.com/Slaveworx/dye.js/blob/master/LICENSE) licensed.

---


CHANGELOG.md
## Version 0.4.3

- Added Gradient Background support (changes color of text elements only)



## FEATURE ROAD MAP
- Add support for text color change bases on background image luminance.
- Add support for the user to input the available colors for the change instead of just black & white.
- Add support for images to change color based on gradient backgrounds.
- Add support for images to change color based on background image luminance. 


dye.ts
/** Dye.js - © 2023 - Tiago M. Galvão - https://deviago.me */

//TYPES AND MODULES DEFINITIONS
module CustomTypes {
  export type attribute = "dye" | "gradient" | "blackimg";
  export type format = "rgb" | "rgba";
}

class Color {
  /**
   * Will check color format and distribute it to the right check color function
   * @param bgColor | string | - Color to be type checked
   * @returns - The new color for the element
   */
  static checkFormat(bgColor: string) {
    if (bgColor.startsWith("rgb(")) return this.checkColor(bgColor, "rgb");
    else if (bgColor.startsWith("rgba("))
      return this.checkColor(bgColor, "rgba");
    else if (bgColor.startsWith("linear-gradient("))
      return this.checkGradient(bgColor);
  }

  /**
   * Will check RGB / RGBA balance
   * @param color | string | - Color to check
   * @param format | string | - Color format, can be rgb or rgba
   * @returns | string | - The new color for the element
   */
  static checkColor(color: string, format: string) {
    let pattern, matches, colorValue, colorOpacity;

    format === "rgb"
      ? (pattern = /^rgb?\((\d+),\s*(\d+),\s*(\d+)(,\s*\d+\.*\d+)?\)$/)
      : (pattern = /^rgba?\((\d+),\s*(\d+),\s*(\d+)(,\s*\d+\.*\d+)?\)$/);

    (matches = color.match(pattern)),
      (colorValue = Math.round(
        (parseInt(matches[1]) * 299 +
          parseInt(matches[2]) * 587 +
          parseInt(matches[3]) * 114) /
          1000
      ));

    colorOpacity = "";
    if (matches[4]) colorOpacity = matches[4].replace(", ", "");

    if (format === "rgb")
      return colorValue > 125 ? "rgb(0,0,0)" : "rgb(255,255,255)";
    else if (format === "rgba")
      return colorValue > 125 || parseFloat(colorOpacity) < 0.5
        ? "rgb(0,0,0)"
        : "rgb(255,255,255)";
  }

  /**
   * Will check gradient brightness
   */
  static checkGradient(color: string) {
    /**
     * Will extract all colors from a gradient string (HEX, RGB OR RGBA) and push them to an array
     * @param gradient | string | Receives the gradient string
     * @returns | array | An array with all extracted colors
     */
    const getGradientColors = (gradient: string): string[] => {
      const colors = gradient.match(/#[0-9a-f]{3,6}|rgba?\([\d\s,./]+\)/gi);
      return colors || [];
    };

    /**
     * Will iterate every color present in the color array to find it's luminance
     * @param color | string |
     * @returns
     */
    const getColorLuminance = (color: string): number => {
      let r, g, b, a;
      if (color.startsWith("#")) {
        // HEX
        r = parseInt(color.substring(1, 3), 16);
        g = parseInt(color.substring(3, 5), 16);
        b = parseInt(color.substring(5, 7), 16);
        a = 1;
      } else if (color.startsWith("rgb")) {
        // RGB or RGBA
        const rgba = color
          .substring(color.indexOf("(") + 1, color.lastIndexOf(")"))
          .split(",")
          .map((c) => parseFloat(c.trim()));
        r = rgba[0];
        g = rgba[1];
        b = rgba[2];
        a = rgba.length > 3 ? rgba[3] : 1;
      } else {
        // Unsupported color format
        return 0;
      }

      // Blend color with white background based on alpha value
      r = Math.round((1 - a) * 255 + a * r);
      g = Math.round((1 - a) * 255 + a * g);
      b = Math.round((1 - a) * 255 + a * b);

      // Calculate luminance from final RGB values
      return (0.299 * r + 0.587 * g + 0.114 * b) / 255;
    };

    //Check the color received
    const gradientMatch = color.match(/^linear-gradient\((.+)\)$/);
    if (!gradientMatch) {
      return;
    }

    const gradientArguments = gradientMatch[1].split(",");

    const gradientColors = getGradientColors(gradientArguments.join(","));

    if (gradientColors.length > 1) {
      //Set defaults if gradient actually qualifies as a gradient.
      const luminances = gradientColors.map(getColorLuminance);

      const minLuminance = Math.min(...luminances);
      // const maxLuminance = Math.max(...luminances);
      let textColor = "#ffffff";

      if (gradientColors.length === 2) {
        textColor = minLuminance > 0.5 ? "#000000" : "#ffffff";
      } else if (gradientColors.length === 3) {
        const middleIndex = Math.floor(gradientColors.length / 2);

        if (middleIndex !== -1) {
          const threshold = minLuminance > 0.9 ? 0.8 : 0.5;
          textColor =
            luminances[middleIndex] > threshold ? "#000000" : "#ffffff";
        }
      }

      return textColor;
    }
  }
}

class Dye extends Color {
  //Class Properties
  public bg: string;
  public imgFilter: string;

  constructor(htmlEntity: any) {
    super();
    //Elements and Props to work with
    let selector = Dye.getProp(htmlEntity, "dye"),
      self = htmlEntity,
      parent = htmlEntity.parentElement,
      isGradient = Dye.getProp(htmlEntity, "gradient"),
      isBlackImg = Dye.getProp(htmlEntity, "blackimg") ? true : false,
      isImg = htmlEntity.tagName === "IMG" ? true : false;

    if (selector === "")
      //Get self bg or in case it's an img, get parent bg
      isGradient
        ? (this.bg = isImg
            ? Dye.getBgGradient(parent)
            : Dye.getBgGradient(self))
        : (this.bg = isImg ? Dye.getBgColor(parent) : Dye.getBgColor(self));
    //If there is indeed a selector passed in the dye parameter, then fetch either bg or gradient bg
    if (typeof selector === "string" && selector !== "")
      isGradient
        ? (this.bg = Dye.getBgGradient(selector))
        : (this.bg = Dye.getBgColor(selector));

    /**
     * Define HTML Element Color
     */
    isImg
      ? Dye.setImgFilter(self, this.bg, isBlackImg)
      : Dye.setColor(self, this.bg);
  }

  /**
   * *Gets Static BG Colors
   * @param element | any | - The element to target or the selector to an element
   * @param selector | boolean - default false | - true if the element is a selector.
   * @returns | string | CSS RGB or RGBA info.
   */
  static getBgColor(element: any) {
    //If receives a selector to query instead of an element
    if (typeof element === "string" && element !== "")
      element = document.querySelector(element);

    //Then get the actual bg color of whatever element was received
    return window.getComputedStyle
      ? window
          .getComputedStyle(element, null)
          .getPropertyValue("background-color")
      : element.style.backgroundColor;
  }

  /**
   * *Get Gradient Backgrounds
   * @param element | any | - The element to target or the selector to an element
   * @param selector | boolean - default false | - true if the element is a selector.
   * @returns | string | CSS Linear Gradient info.
   */
  static getBgGradient(element: any) {
    //If receives a selector to query instead of an element
    if (typeof element === "string" && element !== "")
      element = document.querySelector(element);

    //Then get the actual bg color of whatever element was received
    return window.getComputedStyle
      ? window.getComputedStyle(element, null).getPropertyValue("background")
      : element.style.background;
  }

  /**
   * * Get Props from the Element's attributes object
   * @param element | any | - the element to target
   * @param attribute | atttribute | string | - accepts "dye", "gradient", "blackimg"
   * @returns | string | The attribute value
   */

  static getProp(element: any, attribute: CustomTypes.attribute) {
    let attrs = element.attributes;

    return attribute === "dye"
      ? attrs.getNamedItem(attribute).value
      : attrs.getNamedItem(attribute);
  }

  /**
   * *Will set filter of non img element
   * @param node | any | - The node which color will be changed
   * @param bgColor | string | - The Color to change the element to
   */
  static setColor(node: any, bgColor: string) {
    node.style.color = Color.checkFormat(bgColor);
  }

  /**
   * will set the css filter to change color of Image elements
   * @param img | any | - The image to be targeted
   * @param bgColor | string | - The checked color (not the bg color)
   * @param blackImg | boolean | Returns true if Image is a black image
   */

  static setImgFilter(img: any, bgColor: string, blackImg: boolean) {
    let checkedBg = Color.checkFormat(bgColor),
      filter;
    if (checkedBg === "rgb(0,0,0)") {
      filter = blackImg ? "" : "grayscale(1) invert(1)";
    } else if (checkedBg === "rgb(255,255,255)") {
      filter = blackImg ? "grayscale(1) invert(1)" : "";
    }

    img.style.filter = filter;
    img.style.webkitFilter = filter;
  }
}

//INIT
document.querySelectorAll("[dye]").forEach((htmlEntity) => {
  new Dye(htmlEntity);
});


Classes/classes.js
var Dye = /** @class */ (function () {
    function Dye() {
    }
    /**
     * * Gets Static BG Colors
     * @param element | any | - The element to target or the selector to an element
     * @param selector | boolean - default false | - true if the element is a selector.
     * @returns | string | CSS RGB or RGBA info.
     */
    Dye.getColor = function (element) {
        //If receives a selector to query instead of an element
        if (typeof element === "string")
            element = document.querySelector(element);
        //Then get the actual bg color of whatever element was received
        return window.getComputedStyle
            ? window
                .getComputedStyle(element, null)
                .getPropertyValue("background-color")
            : element.style.backgroundColor;
    };
    /**
     * *Get Gradient Backgrounds
     * @param element | any | - The element to target or the selector to an element
     * @param selector | boolean - default false | - true if the element is a selector.
     * @returns | string | CSS Linear Gradient info.
     */
    Dye.getGradient = function (element) {
        //If receives a selector to query instead of an element
        if (typeof element === "string")
            element = document.querySelector(element);
        //Then get the actual bg color of whatever element was received
        return window.getComputedStyle
            ? window.getComputedStyle(element, null).getPropertyValue("background")
            : element.style.background;
    };
    return Dye;
}());
var Color = /** @class */ (function () {
    function Color() {
    }
    /**
     * will check RGB balance
     */
    Color.checkRGB = function () { };
    /**
     * will check RGBA balance
     */
    Color.checkRGBA = function () { };
    /**
     * Will check gradient brightness
     */
    Color.checkGradient = function () { };
    /**
     * Will set color of non img element
     */
    Color.set = function () { };
    /**
     * will set the color of img elements
     */
    Color.imgSet = function () { };
    return Color;
}());


Classes/classes.ts
class Dye {
  /**
   * * Gets Static BG Colors
   * @param element | any | - The element to target or the selector to an element
   * @param selector | boolean - default false | - true if the element is a selector.
   * @returns | string | CSS RGB or RGBA info.
   */
  static getColor(element: any) {
    //If receives a selector to query instead of an element
    if (typeof element === "string") element = document.querySelector(element);

    //Then get the actual bg color of whatever element was received
    return window.getComputedStyle
      ? window
          .getComputedStyle(element, null)
          .getPropertyValue("background-color")
      : element.style.backgroundColor;
  }

  /**
   * *Get Gradient Backgrounds
   * @param element | any | - The element to target or the selector to an element
   * @param selector | boolean - default false | - true if the element is a selector.
   * @returns | string | CSS Linear Gradient info.
   */
  static getGradient(element: any) {
    //If receives a selector to query instead of an element
    if (typeof element === "string") element = document.querySelector(element);

    //Then get the actual bg color of whatever element was received
    return window.getComputedStyle
      ? window.getComputedStyle(element, null).getPropertyValue("background")
      : element.style.background;
  }
}

class Color {
  /**
   * will check RGB balance
   */
  static checkRGB() {}

  /**
   * will check RGBA balance
   */
  static checkRGBA() {}

  /**
   * Will check gradient brightness
   */
  static checkGradient() {}

  /**
   * Will set color of non img element
   */
  static set() {}

  /**
   * will set the color of img elements
   */

  static imgSet() {}
}


dye.js
/** Dye.js - © 2023 - Tiago M. Galvão - https://deviago.me */
class Color {
    /**
     * Will check color format and distribute it to the right check color function
     * @param bgColor | string | - Color to be type checked
     * @returns - The new color for the element
     */
    static checkFormat(bgColor) {
        if (bgColor.startsWith("rgb("))
            return this.checkColor(bgColor, "rgb");
        else if (bgColor.startsWith("rgba("))
            return this.checkColor(bgColor, "rgba");
        else if (bgColor.startsWith("linear-gradient("))
            return this.checkGradient(bgColor);
    }
    /**
     * Will check RGB / RGBA balance
     * @param color | string | - Color to check
     * @param format | string | - Color format, can be rgb or rgba
     * @returns | string | - The new color for the element
     */
    static checkColor(color, format) {
        let pattern, matches, colorValue, colorOpacity;
        format === "rgb"
            ? (pattern = /^rgb?\((\d+),\s*(\d+),\s*(\d+)(,\s*\d+\.*\d+)?\)$/)
            : (pattern = /^rgba?\((\d+),\s*(\d+),\s*(\d+)(,\s*\d+\.*\d+)?\)$/);
        (matches = color.match(pattern)),
            (colorValue = Math.round((parseInt(matches[1]) * 299 +
                parseInt(matches[2]) * 587 +
                parseInt(matches[3]) * 114) /
                1000));
        colorOpacity = "";
        if (matches[4])
            colorOpacity = matches[4].replace(", ", "");
        if (format === "rgb")
            return colorValue > 125 ? "rgb(0,0,0)" : "rgb(255,255,255)";
        else if (format === "rgba")
            return colorValue > 125 || parseFloat(colorOpacity) < 0.5
                ? "rgb(0,0,0)"
                : "rgb(255,255,255)";
    }
    /**
     * Will check gradient brightness
     */
    static checkGradient(color) {
        /**
         * Will extract all colors from a gradient string (HEX, RGB OR RGBA) and push them to an array
         * @param gradient | string | Receives the gradient string
         * @returns | array | An array with all extracted colors
         */
        const getGradientColors = (gradient) => {
            const colors = gradient.match(/#[0-9a-f]{3,6}|rgba?\([\d\s,./]+\)/gi);
            return colors || [];
        };
        /**
         * Will iterate every color present in the color array to find it's luminance
         * @param color | string |
         * @returns
         */
        const getColorLuminance = (color) => {
            let r, g, b, a;
            if (color.startsWith("#")) {
                // HEX
                r = parseInt(color.substring(1, 3), 16);
                g = parseInt(color.substring(3, 5), 16);
                b = parseInt(color.substring(5, 7), 16);
                a = 1;
            }
            else if (color.startsWith("rgb")) {
                // RGB or RGBA
                const rgba = color
                    .substring(color.indexOf("(") + 1, color.lastIndexOf(")"))
                    .split(",")
                    .map((c) => parseFloat(c.trim()));
                r = rgba[0];
                g = rgba[1];
                b = rgba[2];
                a = rgba.length > 3 ? rgba[3] : 1;
            }
            else {
                // Unsupported color format
                return 0;
            }
            // Blend color with white background based on alpha value
            r = Math.round((1 - a) * 255 + a * r);
            g = Math.round((1 - a) * 255 + a * g);
            b = Math.round((1 - a) * 255 + a * b);
            // Calculate luminance from final RGB values
            return (0.299 * r + 0.587 * g + 0.114 * b) / 255;
        };
        //Check the color received
        const gradientMatch = color.match(/^linear-gradient\((.+)\)$/);
        if (!gradientMatch) {
            return;
        }
        const gradientArguments = gradientMatch[1].split(",");
        const gradientColors = getGradientColors(gradientArguments.join(","));
        if (gradientColors.length > 1) {
            //Set defaults if gradient actually qualifies as a gradient.
            const luminances = gradientColors.map(getColorLuminance);
            const minLuminance = Math.min(...luminances);
            // const maxLuminance = Math.max(...luminances);
            let textColor = "#ffffff";
            if (gradientColors.length === 2) {
                textColor = minLuminance > 0.5 ? "#000000" : "#ffffff";
            }
            else if (gradientColors.length === 3) {
                const middleIndex = Math.floor(gradientColors.length / 2);
                if (middleIndex !== -1) {
                    const threshold = minLuminance > 0.9 ? 0.8 : 0.5;
                    textColor =
                        luminances[middleIndex] > threshold ? "#000000" : "#ffffff";
                }
            }
            return textColor;
        }
    }
}
class Dye extends Color {
    constructor(htmlEntity) {
        super();
        //Elements and Props to work with
        let selector = Dye.getProp(htmlEntity, "dye"), self = htmlEntity, parent = htmlEntity.parentElement, isGradient = Dye.getProp(htmlEntity, "gradient"), isBlackImg = Dye.getProp(htmlEntity, "blackimg") ? true : false, isImg = htmlEntity.tagName === "IMG" ? true : false;
        if (selector === "")
            //Get self bg or in case it's an img, get parent bg
            isGradient
                ? (this.bg = isImg
                    ? Dye.getBgGradient(parent)
                    : Dye.getBgGradient(self))
                : (this.bg = isImg ? Dye.getBgColor(parent) : Dye.getBgColor(self));
        //If there is indeed a selector passed in the dye parameter, then fetch either bg or gradient bg
        if (typeof selector === "string" && selector !== "")
            isGradient
                ? (this.bg = Dye.getBgGradient(selector))
                : (this.bg = Dye.getBgColor(selector));
        /**
         * Define HTML Element Color
         */
        isImg
            ? Dye.setImgFilter(self, this.bg, isBlackImg)
            : Dye.setColor(self, this.bg);
    }
    /**
     * *Gets Static BG Colors
     * @param element | any | - The element to target or the selector to an element
     * @param selector | boolean - default false | - true if the element is a selector.
     * @returns | string | CSS RGB or RGBA info.
     */
    static getBgColor(element) {
        //If receives a selector to query instead of an element
        if (typeof element === "string" && element !== "")
            element = document.querySelector(element);
        //Then get the actual bg color of whatever element was received
        return window.getComputedStyle
            ? window
                .getComputedStyle(element, null)
                .getPropertyValue("background-color")
            : element.style.backgroundColor;
    }
    /**
     * *Get Gradient Backgrounds
     * @param element | any | - The element to target or the selector to an element
     * @param selector | boolean - default false | - true if the element is a selector.
     * @returns | string | CSS Linear Gradient info.
     */
    static getBgGradient(element) {
        //If receives a selector to query instead of an element
        if (typeof element === "string" && element !== "")
            element = document.querySelector(element);
        //Then get the actual bg color of whatever element was received
        return window.getComputedStyle
            ? window.getComputedStyle(element, null).getPropertyValue("background")
            : element.style.background;
    }
    /**
     * * Get Props from the Element's attributes object
     * @param element | any | - the element to target
     * @param attribute | atttribute | string | - accepts "dye", "gradient", "blackimg"
     * @returns | string | The attribute value
     */
    static getProp(element, attribute) {
        let attrs = element.attributes;
        return attribute === "dye"
            ? attrs.getNamedItem(attribute).value
            : attrs.getNamedItem(attribute);
    }
    /**
     * *Will set filter of non img element
     * @param node | any | - The node which color will be changed
     * @param bgColor | string | - The Color to change the element to
     */
    static setColor(node, bgColor) {
        node.style.color = Color.checkFormat(bgColor);
    }
    /**
     * will set the css filter to change color of Image elements
     * @param img | any | - The image to be targeted
     * @param bgColor | string | - The checked color (not the bg color)
     * @param blackImg | boolean | Returns true if Image is a black image
     */
    static setImgFilter(img, bgColor, blackImg) {
        let checkedBg = Color.checkFormat(bgColor), filter;
        if (checkedBg === "rgb(0,0,0)") {
            filter = blackImg ? "" : "grayscale(1) invert(1)";
        }
        else if (checkedBg === "rgb(255,255,255)") {
            filter = blackImg ? "grayscale(1) invert(1)" : "";
        }
        img.style.filter = filter;
        img.style.webkitFilter = filter;
    }
}
//INIT
document.querySelectorAll("[dye]").forEach((htmlEntity) => {
    new Dye(htmlEntity);
});
